src/message.coffee

```diff
@@ -57,45 +62,96 @@ class SlackTextMessage extends TextMessage
+    cursor = 0
+    regex = SlackTextMessage.MESSAGE_REGEX
+    regex.lastIndex = 0
+
+    while (res = regex.exec(text))
+      [m, type, link, label] = res
+
       switch type
         when '@'
-          if label then return "@#{label}"
-          # TODO
-          # user = @dataStore.getUserById link
-          # if user
-          #   return "@#{user.name}"
-          return m
+          if label
+            parts.push(text.slice(cursor, res.index), "@#{label}")
+          else
+            parts.push(text.slice(cursor, res.index), @replaceUser.apply(null, [link, @_client]))
+        
         when '#'
-          if label then return "\##{label}"
-          # TODO
-          # channel = @dataStore.getChannelById link
-          # if channel
-          #   return "\##{channel.name}"
-          return m
+          if label
+            parts.push(text.slice(cursor, res.index), "\##{label}")
+          else
+            parts.push(text.slice(cursor, res.index), @replaceChannel.apply(null, [link, @_client]))
+
         when '!'
           if link in SlackTextMessage.MESSAGE_RESERVED_KEYWORDS
-            return "@#{link}"
+            parts.push(text.slice(cursor, res.index), "@#{link}")
           else if label
-            return label
-          return m
+            parts.push(text.slice(cursor, res.index), label)
+          else
+            parts.push(text.slice(cursor, res.index), m)
+
         else
           link = link.replace /^mailto:/, ''
           if label and -1 == link.indexOf label
-            return "#{label} (#{link})"
+            parts.push(text.slice(cursor, res.index), "#{label} (#{link})")
           else
-            return link
+            parts.push(text.slice(cursor, res.index), link)
+
+      cursor = regex.lastIndex
+      if (res[0].length == 0) 
+        regex.lastIndex++
+
+    parts.push text.slice(cursor)
+
+    return Promise.all(parts).then((substrings) ->
+      return substrings.join('')
+    )
 
-    text = text.replace /&lt;/g, '<'
-    text = text.replace /&gt;/g, '>'
-    text = text.replace /&amp;/g, '&'
+  ###*
+  # Returns name of user with id
+  ###
+  replaceUser: (id, client) ->
+    return client.web.users.info(id).then((user) -> "@#{user.name}")
 
-    if @_channel?.is_im
-      text = "#{@_robot_name} #{text}"     # If this is a DM, pretend it was addressed to us
+  ###*
+  # Returns name of channel with id
+  ###
+  replaceChannel: (id, client) ->
+    return client.web.conversations.info(id).then((channel) -> "\##{channel.name}")
 
-    @text = text
+  ###*
+  # Factory method to construct SlackTextMessage
+  ###
+  @makeSlackTextMessage: (@user, text, rawText, @rawMessage, channel, robot_name, client, cb) ->
+    message = new SlackTextMessage(@user, text, rawText, @rawMessage, channel, robot_name, client)
 
+    if not message.text? then message.buildText(() =>
+      message._client = undefined
+      cb(message)
+    ) else 
+      message._client = undefined
+      cb(message)
```

Comment:
one piece of advice - put this call to `cb()` in a `setImmediate()` so that this function is always deterministically async. the way its written now, depending on the value of `text`, the callback is called either after the function returns or before the function returns, and that can lead to non-deterministic behavior for the caller.